### PR TITLE
[Fix] 'win32' platform (running on Node <12.10.0) - Callback is not a function error

### PR DESCRIPTION
**Pull request by ognjenjevremovic**
✅ Closes: #5492 

# ↪️ Pull Request

Addresses the runtime error `Callback is not a function error` on *win32* platform running on `Node` versions *< 12.10.0*.

## ✔️ PR Todo

- [ ] Added/updated unit tests for this change
- [x] Filled out test instructions (as case there aren't any unit tests)
- [x] Included links to related issues/PRs

## 📝 Additional notes

After the change I ensured that:
 * all tests are passing, by running `npm run test`,
 * no stylistic or unwanted errors are present,
 * formating is consistent.

## 🚨 Manual test instructions

Running the local build of `npkill` tool with the code from the source branch with the included changes in a directory that includes (several) `node_module` (a few directories deep) successfully removes directories with no errors thrown during runtime. The removal is also *asynchronous* which follows the `npkill` convetion.
This change only affects the `win32` platform running `Node` version < 12.10.0.
*No other platforms / Node versions are affected* by this change.

## Additional context of the change (and the bug itself)

I was able to reproduce the issue on `posix` platform by having the `deleteDir` method from `WindowsFilesService` copied over to an appropriate class that extends `UnixFilesService` (based on my OS). In order to get the error thrown during the runtime I had to downgrade my `Node` version to < 12.10.0.

**Why?** The `fs.rmdir` API had a **major change** on the version 12.10.0 (although NOT a breaking change since the new method version is backwards compatible) that allowed the method to support *options* argument.
`v12.10.0 - The recursive, maxBusyTries, and emfileWait options are now supported.`

As taken from the method's typings files the `fs.rmdir` method accepts up to 3 arguments on `Node` versions >= 12.10.0:
```
function rmdir(path: PathLike, callback: NoParamCallback): void;
function rmdir(path: PathLike, options: RmDirOptions, callback: NoParamCallback): void;
```

Prior to `Node` release 12.10.0 `fs.rmdir` method supported only 2 arguments and both of them needed to be provided (some, even older versions of `Node` marked callback parameter as option).
Also, second argument needed to be of type `function`.
```
function rmdir(path: PathLike, callback: NoParamCallback): void;
```

I didn't want to depend on `fs.promises` nor `util.promisify` APIs as that would limit the number of `Node` version supported.
I didn't test the code on versions below `v10.15.3` as per [official releases document found on Nodejs website](https://nodejs.org/en/about/releases/) version `10` is the least supported version of Node with a support ending on Apr 2021 (a few months from the time of writing this).

However, as per current implementation `Node` versions supported are drastically increased since the methods used date to the first very few **production** versions of Node (that I assume people are still running from legacy reasons).

I might be able to supply some unit tests, however I need to *stub* most of the method logic which *imho* might defeat the purpose of the unit test, however I will try to make a commit that will add some sort of unit tests.

### CODE REVIEW

**Review by KonradNojman — APPROVED** [review]
Tested on Node v10.15.3 on Windows (the version where this bug occurred in the first place)

Great job! Works very well ✅ I couldn't find any problem with your changes, so for me they are ready to be merged 💪 (unfortunately cant

**Comment by ognjenjevremovic** [comment]
Thank you for the feedback @KonradNojman and for the kind words.
I'm glad the fix resolves the issue 🙂 .

Let's wait for the @zaldih to go through the changes and approve the PR.
Of course, I'm willing to provide any further changes requested prior to merging.

**Review by zaldih — APPROVED** [review]

**Comment by zaldih** [comment]
Absolutely agree with @KonradNojman . Great work and great research. Surprised with the quality of the PR.

I admire the information gathered about the problem and the proposed solution.

Everything seems to work perfectly and also @NyaGarcia  is doing some testing. As soon as she can give her feedback (a few hours), we merge and release new version!

**Review by NyaGarcia — APPROVED** [review]

**Comment by NyaGarcia** [comment]
I take my hat off to you @ognjenjevremovic, for submitting such a detailed PR. Tested your solution on Windows, with Node v12.0.0 and it works perfectly. Thank you very much helping us out :)

**Comment by KonradNojman** [comment]
Hi @zaldih @NyaGarcia !
Just before merging just want to ask if you had a similar issue to what I wrote in this comment? 
https://github.com/voidcosmos/npkill/issues/87#issuecomment-792323515

Don't want it to be missed and your feedback would be helpful to decide if this is something connected to the PR or maybe because npkill was installed locally or something :) 

Anyway thanks for testing! Glad to see so much progress with this bug fix ❤️

**Comment by zaldih** [comment]
Hey @KonradNojman ! Good point regarding the flicker bug. I have found the cause: now the delete method in windows is synchronous, so the UI will 'freeze' while performing the delete job. Then, when the process finishes, it updates the status to "deleting..." and a few ms later, it changes to "deleted".

I've added a small commit to fix this and refresh the screen before starting the process.

**Comment by ognjenjevremovic** [comment]
@zaldih @NyaGarcia Thank you so much for the kind words! I really do appreciate it ❤️ .
Big props for testing the PR and reporting back.

@KonradNojman thank you for the feedback and the involvement through the whole process. I am really glad we were able to track the issue after some time.
Regarding the flickering, it seems like that @zaldih already addressed the issue in the latest commit attached to the PR.
Would be great if you could pull the latest changes from the branch (head `a2524438`) and test the fix locally once again, when you have time and report back (it would mean a lot!) 🙂 .

Waiting for further feedback; feel free ping me if you require any assistance regarding this and I'll be sure to report back (with a slight delay probably, but no more than a day or two).
Cheers!

**Comment by KonradNojman** [comment]
Hi again @zaldih @ognjenjevremovic ! 

I've did some testing, but the flickering problem still persists ☹️ 

Let me tell you the steps i did to test those changes (unlikely, but maybe I missed something important)
1. pulled the changes from the `a2524438` commit (checked in git logs that i have the newest commit)
2. run `npm install && npm run build` in the local version of the npkill
3. in the test project `npm uninstall npkill` and then `npm install /path/to/the/local/npkill`
4. run the npm script that fires local version of the npkill 

If you have any idea if I can do anything differently, please let me know 🙏 .
And if this procedure is ok, looks like we still have this problem unresolved. Hope we can find and fix it soon 🤞 
Cheers!

**Comment by ognjenjevremovic** [comment]
Thanks for the feedback @KonradNojman. Much appreciated! 🙂 
It seems we might need to give this a closer look.

I am not sure of the impact of this however, perhaps @zaldih or @NyaGarcia are able to prioritize the release based of the flickering issue.
Should we track the flickering as a separate issue?

> Let me tell you the steps i did to test those changes (unlikely, but maybe I missed something important)

That seems the right approach.
As long as you set your HEAD at `a2524438` (by `git checkout a2524438`), prior to building and then followed with install step, I don't see that being an issue.

I might be able to spin up a VM with `win32` platform during the weekend and try to get around it (fingers crossed).
If I'm able to setup the testing environment locally, I'll be glad to give this one a closer look and report back.

**Comment by zaldih** [comment]
You're right. We should prioritize the main problem over the flicker bug. Merge and we launch version! ❤️ 🎉

**Pull request merged**
